Testing/Testing tools.md

```diff
@@ -1,2 +1,268 @@
-Tests in projects help us create stable applications and we should do them as a part of any project. While there are different kinds of tests, we mostly write unit and integration tests in .NET development. Tests are always written in a separate project.
-We use the [xUnit](https://xunit.net/) testing tool by default, along with [Moq](https://github.com/moq/moq4) for mocking and [Fluent Assertions](https://fluentassertions.com/introduction) for building assertion rules. The tests are structured using the Arrange-Act-Assert pattern which helps us write readable and understandable code.
+As it is with a lot of things in software development, when writing tests, we don't want to reinvent the wheel for each project. The .NET ecosystem has a lot of 1st and 3rd party tools available that can make our lives easier when writing tests (and no, we're not thinking of just asking ChatGPT to write them for you). Here are just some of them that we use regularly.
+
+### Testing frameworks
+
+The most popular testing frameworks for .NET are [NUnit](https://nunit.org/), [xUnit](https://xunit.net/) and [MSTest](https://learn.microsoft.com/en-us/dotnet/core/testing/unit-testing-with-mstest). We use xUnit by default because it encourages writing clean code and provides better isolation of tests. Because of that, the following text will mostly concentrate on xUnit features, but the same principles could be applied to different testing frameworks as well.
+
+### Test parameters
+
+When we want to test a piece of code with different parameters, but similar expected results, we could write a single test case for each parameter value. Although this would work, this kind of code duplication should immediately cause a twitch in your eye. There must be a better way to define those tests, right? Test methods are, after all, methods, so why don't we just pass the differences as parameters and let the method define the testing logic around them?
+
+xUnit supports a few ways of defining the test parameters. When a test method is decorated with the `[Theory]` attribute, the test framework knows that there will be more attributes that define the various test parameters.
+
+#### Inline Data
+
+Using the `[InlineData]` attribute is the simplest way to define the test parameters. We repeat the attribute for each set of parameters, and the test framework then knows to repeat that test for each defined set:
+
+``` c#
+[Theory]
+[InlineData("")]
+[InlineData("  ")]
+[InlineData(default(string))]
+public void Validate_WhenNameIsNullOrEmpty_ThenReturnFalse(string name)
+{
+    var user = new AutoFaker<User>()
+        .RuleFor(u => u.Name, _ => name)
+        .Generate();
+
+    var result = _userValidator.Validate(user);
+
+    result.Should().Be(false);
```

Comment:
I may be wrong but I think this line would throw a compilation error. I think that the `Validate()` method doesn't return a boolean directly. Instead, it returns a `ValidationResult` object which has the `IsValid` property. So this line should say `result.IsValid.Should().Be(false)`.

Also, maybe it would be a good idea to use the FluentValidation `Should().BeFalse()` instead of `Should().Be(false)`

Reply:
To be honest, I made up this example validator without having Fluent Validation in mind, I was focusing more on showing how the invalid strings could be tested. But I see your point, the example could be improved by adding something from the "real world" :)

Reply:
Oh I see, not a problem. Looking good now!